src/nbl/ui/CWindowWin32.cpp

```diff
@@ -1,10 +1,13 @@
 #include "nbl/ui/CWindowWin32.h"
+#include <hidusage.h>
 
 #ifdef _NBL_PLATFORM_WINDOWS_
 
 namespace nbl {
 namespace ui
 {
+	static std::multimap<std::pair<core::smart_refctd_ptr<CWindowWin32>, std::string>, IMouseEventChannel> s_mouseEventChannels;
+	static std::multimap<std::pair<core::smart_refctd_ptr<CWindowWin32>, std::string>, IKeyboardEventChannel> s_keyboardEventChannels;
```

Comment:
this is evil.

1) No statics
2) multimap is not thread safe (this will all blow up)
3) a window should hold the storage/ownership of its own channels (at least in the long run)

Reply:
> No statics

why? it's one for all windows and local in translation unit. Fine by me

> multimap is not thread safe (this will all blow up)

Ofc, there will be locks

> a window should hold the storage/ownership of its own channels (at least in the long run)

Havent thought about it this way. We have callbacks for input device connected/disconnected so that user knows which channel is for which window at the very moment of acquiring the channel. 
At the end of the day we always end up with a map, either native_window_handle->CWindowWin32 (and window holds all of its channels) or window->channels multimap

Reply:
@sadiuk 
* What is the string in the key?
* Use `nbl::core::multimap` instead
* Value type should be smart_refctd_ptr
* You need a way of finding CWindowWin32 object from HWND (thats what you'll get from win32 event), so maybe storing channels inside window objects is in fact a better option (then you only need HWND->CWindowWin32 map)

Reply:
if you store channel inside a window, then you wouldn't need a HWND map though?

Reply:
> * What is the string in the key?

That's  a device name. That's the most unique thing i found to identify connected input devices

>  You need a way of finding CWindowWin32 object from HWND (thats what you'll get from win32 event), so maybe storing channels inside window objects is in fact a better option (then you only need HWND->CWindowWin32 map)

You can see i call `SetWindowLongPtr()` with a `CWindowWin32*` when initializing the window, you can bind any user param with `HWND`, so no need for map here, but yeah, i guess making it a member of CWindowWin32 is better.

Reply:
> That's a device name. That's the most unique thing i found to identify connected input devices

ok, so there's no other thing to let me decide which channel (mouse/KB/etc.) to push an event to when I "wake up" in the window event callback.

We have to look up the channel "by string"?

Reply:
> We have to look up the channel "by string"?

nvm, i'll use device `HANDLE`s instead, missed that.

Reply:
You'll still be looking up a channel based on some key.

Reply:
well it depends on how big is the set of things you'll look through and what is the key.

A key which is a string is bad, because it takes long to compare plus there's always a string to construct/destruct (memory allocation) temporarily.

Then if you use an integer or pointer/handle, there's a choice of data structure to consider:
Vector : for very few elements (n<32) its actually faster to insert and erase (especially if you wont be frequently adding or removing elements, like in the case of connecting and disconnecting input devices) in O(n) and do a O(n) linear search when looking up because of cache effects
Map : for few elements (n<1000) a map (which is usually implemented as a red-black tree) is faster than a hashmap (especially for a fat key) despite its O(nlogn) insertion, deletion and searching. This is because a hashmap does have overheads and the hash function scatters your elements all over the place, this is a worst case scenario for CPU caches.
HashMap (unordered_map) : for many elements (n>1000), theoretically O(1) but when you dont have enough elements, the overheads are large.